### PR TITLE
REST api calls

### PR DESCRIPTION
**Pull request by zaucker**
Add API routes to

- run a simulation (optionally specifying a technical file within the model)
- get a technical file (as reference)
- get a LaTeX docu (as reference)

and some curl scripts for testing.

### CODE REVIEW

File: lib/Agrammon/DataSource/CSV.pm6 [hunk 1]
```diff
@@ -44,6 +44,27 @@ class Agrammon::DataSource::CSV {
         }
     }
 
+    method from-csv($simulation-name, $dataset-id, $csv-data) {
+        my @group = $csv-data.split("\n");
+        my $input = Agrammon::Inputs.new(:$simulation-name, :$dataset-id);
+        for @group.map(*.split(';')) {
```
Comment (jnthn) [inline lib/Agrammon/DataSource/CSV.pm6:50]: It's maybe worth avoiding the temporary array here and just inlining the `$csv-data.split("\n")` in place of the group. (Also, it saves me complaining that `@group` is a slightly odd name for one record/line in the CSV file).

File: lib/Agrammon/Web/APIRoutes.pm6 [hunk 2]
```diff
@@ -28,14 +26,64 @@ my class AgrammonAPITokenMiddleware does Cro::APIToken::Middleware {
...
-        }
+        include rest-api-routes($schema, $ws);
     }
 }
 
+sub rest-api-routes (Str $schema, Agrammon::Web::Service $ws) is export {
+    openapi $schema.IO, {
+        operation 'greetUser', -> APIUser $user {
+            content 'application/json', { message => "Hello $user.firstname()" }
+        }
```
Comment (jnthn) [inline lib/Agrammon/Web/APIRoutes.pm6:52]: Do we really want to keep this?

File: lib/Agrammon/TechnicalParser.pm6 [hunk 3]
```diff
@@ -26,3 +27,13 @@ sub parse-technical(Str $to-parse) is export {
         die "Failed to parse technical configuration: unknown error";
     }
 }
+
+sub load-technical(IO::Path $model-path, Str $technical-file) is export {
+    my $tech-input = $model-path.IO.add($technical-file // 'technical.cfg');
```
Comment (jnthn) [inline lib/Agrammon/TechnicalParser.pm6:32]: Note that `add` will interpret any `..` sequences, so there's a potential vulnerability here given we're letting the caller of the API pick the path.

File: lib/Agrammon/Web/APIRoutes.pm6 [hunk 4]
```diff
@@ -28,14 +26,64 @@ my class AgrammonAPITokenMiddleware does Cro::APIToken::Middleware {
...
+        operation 'getLatex', -> APIUser $user, :$technical = 'technical.cfg', :$sort = 'model' {
+            content 'text/plain', $ws.get-latex($technical, $sort)
```
Comment (jnthn) [inline lib/Agrammon/Web/APIRoutes.pm6:54]: Possible directory traversal vuln via `$technical`

File: lib/Agrammon/Web/APIRoutes.pm6 [hunk 5]
```diff
@@ -28,14 +26,64 @@ my class AgrammonAPITokenMiddleware does Cro::APIToken::Middleware {
...
+        }
 
+        operation 'getTechnical', -> APIUser $user, :$technical = 'technical.cfg' {
+            content 'text/plain', $ws.get-technical($technical);
```
Comment (jnthn) [inline lib/Agrammon/Web/APIRoutes.pm6:58]: Possible directory traversal vuln via `$technical`

File: lib/Agrammon/Web/APIRoutes.pm6 [hunk 6]
```diff
@@ -28,14 +26,64 @@ my class AgrammonAPITokenMiddleware does Cro::APIToken::Middleware {
...
+        operation 'runSimulation', -> APIUser $user {
+            request-body 'multipart/form-data' => -> (
+                :$simulation!,  :$dataset!, :$inputs!, :$technical='',
+                :$model = 'version6', :$variants = 'Base', :$language = 'de', :$format = 'text/plain',
+                :$print-only = '', :$include-filters = 'false', :$all-filters = 'false'
+            ) {
+                my $type = $inputs.content-type;
+                if $type ne 'text/csv' {
+                    my $error = "Content type is '$type', must be 'text/csv'";
+                    note $error;
```
Comment (jnthn) [inline lib/Agrammon/Web/APIRoutes.pm6:70]: Wanted logging or leftover?

File: lib/Agrammon/Web/APIRoutes.pm6 [hunk 7]
```diff
@@ -28,14 +26,64 @@ my class AgrammonAPITokenMiddleware does Cro::APIToken::Middleware {
...
+                    bad-request 'application/json', %( error => $error );
```
Comment (jnthn) [inline lib/Agrammon/Web/APIRoutes.pm6:71]: Could be `:$error`

File: lib/Agrammon/Web/Service.pm6 [hunk 8]
```diff
@@ -198,10 +203,88 @@ class Agrammon::Web::Service {
...
+        :$model-version, :$variants, :$technical-file,
+        :$language, :$format, :$print-only,
+        :$include-filters, :$all-filters
+    ) {
+        my $input = Agrammon::DataSource::CSV.new.from-csv($simulation-name, $dataset-name, $csv-data);
+
+        my $model;
+        if $model-version {
+            my $model-top   = $!cfg.model-top;
+            my $model-path  = self.model.path.IO.parent.add($model-version).add($model-top);
```
Comment (jnthn) [inline lib/Agrammon/Web/Service.pm6:239]: Directory traversal vuln here also

File: lib/Agrammon/Web/APIRoutes.pm6 [hunk 9]
```diff
@@ -28,14 +26,64 @@ my class AgrammonAPITokenMiddleware does Cro::APIToken::Middleware {
...
+                }
+                else {
+                    my $input-data = $inputs.body-text;
+                    my $results = $ws.get-outputs-from-csv(
+                        $user,
+                        ~$simulation, ~$dataset, ~$input-data,
+                        :model-version(~$model), :variants(~$variants), :technical-file(~$technical),
+                        :language(~$language), :format(~$format), :print-only(~$print-only),
```
Comment (jnthn) [inline lib/Agrammon/Web/APIRoutes.pm6:79]: If we were being really well behaved here in terms of HTTP we'd take the format in the `Accept` header.

File: t/apiroutes.t [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+                    :simulation('Test'),
+                    :dataset('Hochrechnung'),
+                    Cro::HTTP::Body::MultiPartFormData::Part.new(
+                        headers => [Cro::HTTP::Header.new(
+                            :name('Content-type'),
+                            :value('text/csv')
+                        )],
+                        :name('inputs'),
+                        :filename('test.csv'),
+#                        :body-text('some;test;data'.encode)
```
Comment (jnthn) [inline t/apiroutes.t:80]: Leftover experimenting, or hopeful future improvement? :-)

**Review by jnthn — COMMENTED** [review]
The security issues mentioned can be addressed by using https://github.com/raku-community-modules/IO-Path-ChildSecure

File: lib/Agrammon/DataSource/CSV.pm6 (hunk 1 above)
Comment (zaucker) [reply lib/Agrammon/DataSource/CSV.pm6:50]: :-) ... done.

File: lib/Agrammon/Web/APIRoutes.pm6 (hunk 2 above)
Comment (zaucker) [reply lib/Agrammon/Web/APIRoutes.pm6:52]: I found it a cute test route, but you're right. Removed.

File: lib/Agrammon/Web/APIRoutes.pm6 (hunk 6 above)
Comment (zaucker) [reply lib/Agrammon/Web/APIRoutes.pm6:70]: Wanted logging, but nobody is really going to look at it, so removed.

File: lib/Agrammon/Web/APIRoutes.pm6 (hunk 7 above)
Comment (zaucker) [reply lib/Agrammon/Web/APIRoutes.pm6:71]: Done.

File: t/apiroutes.t (hunk 10 above)
Comment (zaucker) [reply t/apiroutes.t:80]: Experimenting, removed.

File: lib/Agrammon/Web/APIRoutes.pm6 (hunk 9 above)
Comment (zaucker) [reply lib/Agrammon/Web/APIRoutes.pm6:79]: Makes sense, done.

File: lib/Agrammon/TechnicalParser.pm6 (hunk 3 above)
Comment (zaucker) [reply lib/Agrammon/TechnicalParser.pm6:32]: Fixed

File: lib/Agrammon/Web/APIRoutes.pm6 (hunk 4 above)
Comment (zaucker) [reply lib/Agrammon/Web/APIRoutes.pm6:54]: Used via `TechnicalParser::load-technical()` which uses `child-secure`.

File: lib/Agrammon/Web/APIRoutes.pm6 (hunk 5 above)
Comment (zaucker) [reply lib/Agrammon/Web/APIRoutes.pm6:58]: Used via `TechnicalParser::load-technical()` which uses `child-secure`.

File: share/agrammon-rest.openapi [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+                model:
+                    type: string
+                variants:
+                    type: string
+                print-only:
+                    type: string
+                language:
+                    type: string
+                format:
+                    type: string
```
Comment (jnthn) [inline share/agrammon-rest.openapi:122]: format went away in favor of the accept header, so should be removed here also

Comment (zaucker) [reply share/agrammon-rest.openapi:122]: Done.

**Review by jnthn — APPROVED** [review]

**Pull request merged**